web/src/app/webview/src/contextManager.ts

```diff
@@ -51,13 +51,50 @@ export default class ContextManager extends ContextManagerBase {
     return this._activeKeyboard;
   }
 
-  set activeKeyboard(kbd: {keyboard: Keyboard, metadata: KeyboardStub}) {
-    const priorEntry = this._activeKeyboard;
-    this._activeKeyboard = kbd;
+  setKeyboardActiveForTarget(kbd: {keyboard: Keyboard, metadata: KeyboardStub}, target: OutputTarget) {
+    // `target` is irrelevant for `app/webview`, as it'll only ever use 'global' keyboard settings.
 
-    if(priorEntry.keyboard != kbd.keyboard || priorEntry.metadata != kbd.metadata) {
-      this.emit('keyboardchange', kbd);
-      this.resetContext();
+    // Clone the object to prevent accidental by-reference changes.
+    this._activeKeyboard = {...kbd};
+  }
+
+  /**
+   * Reflects the active 'target' upon which any `set activeKeyboard` operation will take place.
+   * For app/webview... there's only one target, thus only a "global default" matters.
+   */
+  protected get keyboardTarget(): Mock {
+    return null;
+  }
+
+
+  public async activateKeyboard(keyboardId: string, languageCode?: string, saveCookie?: boolean): Promise<boolean> {
+    try {
+      return await super.activateKeyboard(keyboardId, languageCode, saveCookie);
+    } catch(err) {
+      // Fallback behavior - we're embedded in a touch-device's webview, so we need to keep a keyboard visible.
+      const defaultStub = this.keyboardCache.defaultStub;
+      await this.activateKeyboard(defaultStub.id, defaultStub.langId, true).catch(() => {});
+
+      throw err; // since the consumer may want to do its own error-handling.
     }
   }
+
+  protected prepareKeyboardForActivation(
+    keyboardId: string,
+    languageCode?: string
+  ): {keyboard: Promise<Keyboard>, metadata: KeyboardStub} {
+    const originalKeyboard = this.activeKeyboard;
+    const activatingKeyboard = super.prepareKeyboardForActivation(keyboardId, languageCode);
+
+    // Probably isn't necessary at this point - osk.refreshLayout() exists - but
+    // it's best to keep it around for now and verify later.
+    if(originalKeyboard.metadata.id == activatingKeyboard.metadata.id) {
+      activatingKeyboard.keyboard = activatingKeyboard.keyboard.then((kbd) => {
```

Comment:
Is this correct? `activatingKeyboard.keyboard` gets overwritten with a promise?

Reply:
Note:  activat**ing**.  The keyboard is not yet activated - and quite possibly because it has not yet been loaded.  Thus, that field's typing is for `Promise<Keyboard>`, not `Keyboard`.

https://github.com/keymanapp/keyman/blob/3c185b31826864364d69999de708ebd993a853f3/web/src/engine/main/src/contextManagerBase.ts#L278-L281

So... "yes", except that we're less "overwriting" the `Promise` for the loading keyboard and more "extending" / "filtering" the `Promise`.